aiida/backends/tests/engine/test_calc_job.py

```diff
@@ -93,6 +93,25 @@ def prepare_for_submission(self, folder):
         with self.assertRaises(TypeError):
             launch.run(SimpleCalcJob, code=self.code, metadata={'options': orm.Dict(dict={'a': 1})})
 
+    def test_computer(self):
+        """Test passing an explicit `computer` in the `metadata`."""
+        inputs = {
+            'code': self.code,
```

Comment:
Looking here https://github.com/aiidateam/aiida-core/pull/3400/files#diff-4faa11e49a5f253a316f6d5f1e156cccR31
this seems to be a code with a remote executable.

So, I'm wondering: Shouldn't this test actually fail?
And shouldn't we have another test with a "local" code?
So, what is this test actually testing?

Reply:
It is just testing that the process spec validation, which is triggered by constructing the process instance, does not except, which used to be the case when specifying the `computer` in `metadata.options`, even though it had a corresponding port. None of the tests in the module actually "run" the calculation. Once we move to `pytest` we could take the fixtures I wrote for `aiida-quantumespresso` to easily check more, such as the content of the input files written by `prepare_for_submission`.

Also regarding local and remote codes: there are no real restrictions on this. The only check is that when the `CalcJobNode` is stored, `_validate` will check that _a_ computer has been set. If a remote code is passed in the inputs, its corresponding computer is automatically set. For a local code, one has to do this manually.

Reply:
> Also regarding local and remote codes: there are no real restrictions on this.

I see. Would it make sense to add one, though?
I think we should warn when people are setting a computer for a remote code since this likely means they're using aiida incorrectly/

Reply:
I agree, but I was taking care of this in a separate PR. A user reported an issue in #3395 that was most likely him accidentally using a local code. I told him how to fix it but kept the issue open to improve the validation. I think instead of the validation on the node level, it should be done on the process spec.

Reply:
Ok, that's fine of course.
We should just keep in mind that this test here should fail once this extra check is implemented